File: android/Showcase/src/main/java/com/spruceid/mobilesdkexample/verifier/VerifyKioskView.kt
package com.spruceid.mobilesdkexample.verifier

import android.bluetooth.BluetoothAdapter
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.border
import androidx.compose.foundation.Image
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.offset
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.ui.unit.IntOffset
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.layout.onGloballyPositioned
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.unit.sp
import com.spruceid.mobilesdkexample.R
import androidx.navigation.NavController
import com.google.accompanist.permissions.ExperimentalPermissionsApi
import com.spruceid.mobile.sdk.BLESessionStateDelegate
import com.spruceid.mobile.sdk.IsoMdlReader
import com.spruceid.mobile.sdk.getBluetoothManager
import com.spruceid.mobile.sdk.getPermissions
import com.spruceid.mobile.sdk.ui.GenericCameraXScanner
import com.spruceid.mobile.sdk.ui.QrCodeAnalyzer
import com.spruceid.mobilesdkexample.LoadingView
import com.spruceid.mobilesdkexample.navigation.Screen
import com.spruceid.mobilesdkexample.ui.theme.ColorBase1
import com.spruceid.mobilesdkexample.ui.theme.ColorBase50
import com.spruceid.mobilesdkexample.ui.theme.ColorBlue100
import com.spruceid.mobilesdkexample.ui.theme.ColorBlue200
import com.spruceid.mobilesdkexample.ui.theme.ColorBlue300
import com.spruceid.mobilesdkexample.ui.theme.ColorBlue400
import com.spruceid.mobilesdkexample.ui.theme.ColorBlue600
import com.spruceid.mobilesdkexample.ui.theme.ColorBlue800
import com.spruceid.mobilesdkexample.ui.theme.ColorRose50
import com.spruceid.mobilesdkexample.ui.theme.ColorRose200
import com.spruceid.mobilesdkexample.ui.theme.ColorRose300
import com.spruceid.mobilesdkexample.ui.theme.ColorRose600
import com.spruceid.mobilesdkexample.ui.theme.ColorRose700
import com.spruceid.mobilesdkexample.ui.theme.ColorStone300
import com.spruceid.mobilesdkexample.ui.theme.ColorStone600
import com.spruceid.mobilesdkexample.ui.theme.ColorStone950
import com.spruceid.mobilesdkexample.ui.theme.Inter
import com.spruceid.mobilesdkexample.utils.Toast
import com.spruceid.mobilesdkexample.utils.activityHiltViewModel
import com.spruceid.mobilesdkexample.utils.checkAndRequestBluetoothPermissions
import com.spruceid.mobilesdkexample.viewmodels.TrustedCertificatesViewModel
import kotlinx.coroutines.launch

val kioskElements: Map<String, Map<String, Boolean>> = mapOf(
    "org.iso.18013.5.1" to mapOf(
        "family_name" to false,
        "given_name" to false,
        "birth_date" to false,
        "issue_date" to false,
        "expiry_date" to false,
    )
)

enum class KioskState {
    SCANNING,
    CONNECTING,
    SUCCESS,
    ERROR,
    BLUETOOTH_REQUIRED
}

@OptIn(ExperimentalPermissionsApi::class)
@Composable
fun VerifyKioskView(navController: NavController) {
    val trustedCertificatesViewModel: TrustedCertificatesViewModel = activityHiltViewModel()
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    var reader: IsoMdlReader? = null

    var kioskState by remember { mutableStateOf(KioskState.SCANNING) }
    var scannedUri by remember { mutableStateOf<String?>(null) }
    var isBluetoothEnabled by remember {
        mutableStateOf(getBluetoothManager(context)?.adapter?.isEnabled ?: false)
    }

    val launcherMultiplePermissions = rememberLauncherForActivityResult(
        ActivityResultContracts.RequestMultiplePermissions()
    ) { permissionsMap ->
        permissionsMap.values.reduce { acc, next -> acc && next }
    }

    // Bluetooth state listener
    DisposableEffect(Unit) {
        val receiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context?, intent: Intent?) {
                if (intent?.action == BluetoothAdapter.ACTION_STATE_CHANGED) {
                    val state = intent.getIntExtra(BluetoothAdapter.EXTRA_STATE, BluetoothAdapter.ERROR)
                    when (state) {
                        BluetoothAdapter.STATE_OFF -> isBluetoothEnabled = false
                        BluetoothAdapter.STATE_ON -> isBluetoothEnabled = true
                    }
                }
            }
        }
        context.registerReceiver(receiver, IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED))
        onDispose { context.unregisterReceiver(receiver) }
    }

    val bleCallback: BLESessionStateDelegate = object : BLESessionStateDelegate() {
        override fun update(state: Map<String, Any>) {
            if (state.containsKey("mdl")) {
                reader?.handleMdlReaderResponseData(state["mdl"] as ByteArray)
                kioskState = KioskState.SUCCESS
            }
        }

        override fun error(error: Exception) {
            kioskState = KioskState.ERROR
        }
    }

    // Auto-retry when Bluetooth enabled
    LaunchedEffect(isBluetoothEnabled) {
        if (isBluetoothEnabled && kioskState == KioskState.BLUETOOTH_REQUIRED && scannedUri != null) {
            val bluetooth = getBluetoothManager(context.applicationContext)
            kioskState = KioskState.CONNECTING
            try {
                reader = IsoMdlReader(
                    bleCallback,
                    scannedUri!!,
                    kioskElements,
                    trustedCertificatesViewModel.trustedCertificates.value.map { it.content },
                    bluetooth!!,
                    context.applicationContext
                )
            } catch (e: Exception) {
                e.localizedMessage?.let { Toast.showError(it) }
                kioskState = KioskState.ERROR
            }
        }
    }

    fun back() {
        navController.navigate(Screen.HomeScreen.route.replace("{tab}", "verifier")) {
            popUpTo(0)
        }
    }

    fun startOver() {
        scannedUri = null
        reader = null
        kioskState = KioskState.SCANNING
    }

    fun onRead(content: String) {
        scannedUri = content

        checkAndRequestBluetoothPermissions(
            context.applicationContext,
            getPermissions().toTypedArray(),
            launcherMultiplePermissions
        )

        val bluetooth = getBluetoothManager(context.applicationContext)

        if (bluetooth?.adapter?.isEnabled != true) {
            kioskState = KioskState.BLUETOOTH_REQUIRED
            return
        }

        kioskState = KioskState.CONNECTING

        scope.launch {
            try {
                reader = IsoMdlReader(
                    bleCallback,
                    content,
                    kioskElements,
                    trustedCertificatesViewModel.trustedCertificates.value.map { it.content },
                    bluetooth,
                    context.applicationContext
                )
            } catch (e: Exception) {
                e.localizedMessage?.let { Toast.showError(it) }
                kioskState = KioskState.ERROR
            }
        }
    }

    when (kioskState) {
        KioskState.SCANNING -> {
            KioskScanView(onCancel = ::back, onRead = ::onRead)
        }
        KioskState.CONNECTING -> {
            LoadingView(
                loadingText = "Connecting...",
                cancelButtonLabel = "Cancel",
                onCancel = ::back
            )
        }
        KioskState.SUCCESS -> {
            KioskSuccessView(onStartOver = ::startOver)
        }
        KioskState.ERROR -> {
            KioskFailureView(onStartOver = ::startOver)
        }
        KioskState.BLUETOOTH_REQUIRED -> {
            Column(
                modifier = Modifier.fillMaxSize(),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.Center
            ) {
                Text("Turn on Bluetooth to continue.")
            }
        }
    }
}

@Composable
fun KioskScanView(
    onCancel: () -> Unit,
    onRead: (String) -> Unit
) {
    val scannerWidthMultiplier = 0.6f
    val cornerRadiusDp = 12.dp

    var scannerBottomPx by remember { mutableFloatStateOf(0f) }
    var badgeHeight by remember { mutableFloatStateOf(0f) }

    GenericCameraXScanner(
        title = "",
        subtitle = "",
        onCancel = onCancel,
        hideCancelButton = true,
        fontFamily = Inter,
        imageAnalyzer = QrCodeAnalyzer(
            onQrCodeScanned = { result -> onRead(result) }
        ),
        background = {
            Box(modifier = Modifier.fillMaxSize()) {
                // Background
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .background(ColorBase1)
                        .drawWithContent {
                            val canvasWidth = size.width
                            val canvasHeight = size.height
                            val scannerWidth = canvasWidth * scannerWidthMultiplier
                            val left = (canvasWidth - scannerWidth) / 2
                            val top = canvasHeight * 0.35f
                            val cornerRadius = cornerRadiusDp.toPx()

                            scannerBottomPx = top + scannerWidth

                            drawContent()

                            // Camera cut
                            drawRoundRect(
                                topLeft = Offset(left, top),
                                size = Size(scannerWidth, scannerWidth),
                                color = Color.Transparent,
                                blendMode = BlendMode.SrcIn,
                                cornerRadius = CornerRadius(cornerRadius),
                            )

                            // Border
                            drawRoundRect(
                                topLeft = Offset(left, top),
                                size = Size(scannerWidth, scannerWidth),
                                color = ColorBlue600,
                                style = Stroke(4.dp.toPx()),
                                cornerRadius = CornerRadius(cornerRadius),
                            )
                        }
                )

                // Header
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .statusBarsPadding()
                        .padding(horizontal = 30.dp)
                        .padding(top = 20.dp)
                ) {
                    KioskHeader()
                }

                // Title
                Column(
                    modifier = Modifier
                        .fillMaxWidth()
                        .statusBarsPadding()
                        .padding(top = 170.dp),
                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    Text(
                        text = "Scan QR Code",
                        fontFamily = Inter,
                        fontWeight = FontWeight.Bold,
                        fontSize = 28.sp,
                        color = ColorBlue600
                    )

                    Spacer(modifier = Modifier.height(8.dp))

                    Text(
                        text = "Present your digital ID QR code",
                        fontFamily = Inter,
                        fontWeight = FontWeight.Normal,
                        fontSize = 16.sp,
                        color = ColorStone600
                    )
                }

                // Scanning badge
                Box(
                    modifier = Modifier
                        .align(Alignment.TopCenter)
                        .offset {
                            IntOffset(0, (scannerBottomPx - (badgeHeight / 2) - 10.dp.toPx()).toInt())
                        }
                        .background(ColorBlue600, RoundedCornerShape(100.dp))
                        .padding(horizontal = 16.dp, vertical = 8.dp)
                        .onGloballyPositioned { coordinates ->
                            badgeHeight = coordinates.size.height.toFloat()
                        },
                    contentAlignment = Alignment.Center
                ) {
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(8.dp)
                    ) {
                        CircularProgressIndicator(
                            modifier = Modifier.size(16.dp),
                            color = Color.White,
                            strokeWidth = 2.dp
                        )
                        Text(
                            text = "Detecting...",
                            fontFamily = Inter,
                            fontWeight = FontWeight.Medium,
                            fontSize = 14.sp,
                            color = Color.White
                        )
                    }
                }

                // Cancel button
                Column(
                    modifier = Modifier
                        .fillMaxSize()
                        .navigationBarsPadding()
                        .padding(horizontal = 30.dp)
                        .padding(bottom = 30.dp),
                    verticalArrangement = Arrangement.Bottom,
                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    Button(
                        onClick = onCancel,
                        shape = RoundedCornerShape(8.dp),
                        colors = ButtonDefaults.buttonColors(
                            containerColor = Color.Transparent,
                            contentColor = Color.Black,
                        ),
                        border = BorderStroke(1.dp, ColorStone300),
                        modifier = Modifier.fillMaxWidth()
                    ) {
                        Text(
                            text = "Cancel",
                            fontFamily = Inter,
                            fontWeight = FontWeight.SemiBold,
                            color = Color.Black,
                        )
                    }
                }
            }
        }
    )
}

@Composable
fun KioskHeader() {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(top = 15.dp),
        horizontalArrangement = Arrangement.Center,
        verticalAlignment = Alignment.CenterVertically
    ) {
        Image(
            painter = painterResource(id = R.drawable.spruce_logo),
            contentDescription = "Spruce Logo",
            modifier = Modifier.size(28.dp)
        )
        Spacer(modifier = Modifier.width(8.dp))
        Text(
            text = "Spruce County",
            fontFamily = Inter,
            fontWeight = FontWeight.Bold,
            fontSize = 24.sp,
            color = ColorStone950
        )
    }
}

@Composable
fun KioskSuccessView(onStartOver: () -> Unit) {
    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(ColorBase1)
            .clickable { onStartOver() }
    ) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(30.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            KioskHeader()

            Spacer(modifier = Modifier.weight(1f))

            // Icon
            Box(
                modifier = Modifier
                    .size(80.dp)
                    .shadow(4.dp, CircleShape)
                    .background(
                        brush = Brush.verticalGradient(listOf(ColorBase50, ColorBlue200)),
                        shape = CircleShape
                    )
                    .border(2.dp, ColorBlue300, CircleShape),
                contentAlignment = Alignment.Center
            ) {
                Icon(
                    imageVector = Icons.Default.Check,
                    contentDescription = "Success",
                    modifier = Modifier.size(40.dp),
                    tint = ColorBlue600
                )
            }

            Spacer(modifier = Modifier.height(24.dp))

            Text(
                text = "Welcome!",
                fontFamily = Inter,
                fontWeight = FontWeight.Bold,
                fontSize = 36.sp,
                color = ColorStone950
            )

            Spacer(modifier = Modifier.height(24.dp))

            Text(
                text = "Your identity has been successfully verified using your mobile driver's license, and a confirmation has been sent to your email.",
                fontFamily = Inter,
                fontSize = 18.sp,
                color = ColorStone600,
                textAlign = TextAlign.Center,
                modifier = Modifier.padding(horizontal = 20.dp)
            )

            Spacer(modifier = Modifier.weight(1f))
        }
    }
}

@Composable
fun KioskFailureView(onStartOver: () -> Unit) {
    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(ColorBase1)
    ) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(30.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            KioskHeader()

            Spacer(modifier = Modifier.weight(1f))

            // Icon
            Box(
                modifier = Modifier
                    .size(80.dp)
                    .shadow(4.dp, CircleShape)
                    .background(
                        brush = Brush.verticalGradient(listOf(ColorRose50, ColorRose200)),
                        shape = CircleShape
                    )
                    .border(2.dp, ColorRose300, CircleShape),
                contentAlignment = Alignment.Center
            ) {
                Icon(
                    imageVector = Icons.Default.Close,
                    contentDescription = "Invalid",
                    modifier = Modifier.size(40.dp),
                    tint = ColorRose600
                )
            }

            Spacer(modifier = Modifier.height(24.dp))

            Text(
                text = "Invalid",
                fontFamily = Inter,
                fontWeight = FontWeight.Bold,
                fontSize = 36.sp,
                color = ColorRose700
            )

            Spacer(modifier = Modifier.height(24.dp))

            Text(
                text = "Your ID has not been accepted. Please try again or check in with reception for verification.",
                fontFamily = Inter,
                fontSize = 18.sp,
                color = ColorStone600,
                textAlign = TextAlign.Center,
                modifier = Modifier.padding(horizontal = 20.dp)
            )

            Spacer(modifier = Modifier.weight(1f))

            Button(
                onClick = onStartOver,
                shape = RoundedCornerShape(30.dp),
                colors = ButtonDefaults.buttonColors(
                    containerColor = ColorBlue100,
                    contentColor = ColorBlue800
                ),
                border = BorderStroke(1.dp, ColorBlue400),
                contentPadding = PaddingValues(horizontal = 40.dp, vertical = 14.dp)
            ) {
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    Icon(
                        imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                        contentDescription = "Back",
                        modifier = Modifier.size(20.dp)
                    )
                    Text(
                        text = "Start over",
                        fontFamily = Inter,
                        fontWeight = FontWeight.Medium,
                        fontSize = 16.sp
                    )
                }
            }

            Spacer(modifier = Modifier.height(60.dp))
        }
    }
}

File: ios/Showcase/Targets/AppUIKit/Sources/verifier/VerifyKioskView.swift
import AVFoundation
import SpruceIDMobileSdk
import SwiftUI

struct VerifyKiosk: Hashable {}

let kioskElements = [
    "org.iso.18013.5.1": [
        "family_name": false,
        "given_name": false,
        "birth_date": false,
        "issue_date": false,
        "expiry_date": false,
    ]
]

public struct VerifyKioskView: View {
    @Binding var path: NavigationPath
    @State private var scanned: String?

    var trustedCertificates = TrustedCertificatesDataStore.shared.getAllCertificates()

    public var body: some View {
        if scanned == nil {
            KioskScanView(
                onCancel: onCancel,
                onRead: { code in self.scanned = code }
            )
        } else {
            KioskReaderView(
                uri: scanned!,
                requestedItems: kioskElements,
                trustAnchorRegistry: trustedCertificates.map { $0.content },
                onCancel: onCancel,
                onStartOver: { self.scanned = nil },
                path: $path
            )
        }
    }

    func onCancel() {
        self.scanned = nil
        path.removeLast()
    }
}

// MARK: - Scan View

struct KioskScanView: View {
    var onCancel: () -> Void
    var onRead: (String) -> Void

    @State private var cameraPermission: Permission = .idle

    private var scannerSize: CGFloat {
        UIScreen.main.bounds.width * 0.6
    }

    var body: some View {
        VStack(spacing: 0) {
            KioskHeader()

            Spacer().frame(height: 100)

            // Title
            VStack(spacing: 0) {
                Text("Scan QR Code")
                    .font(.system(size: 28, weight: .bold))
                    .foregroundColor(Color("ColorBlue600"))

                Spacer().frame(height: 8)

                Text("Present your digital ID QR code")
                    .font(.system(size: 16))
                    .foregroundColor(Color("ColorStone600"))

                Spacer().frame(height: 24)

                // Scanner
                ZStack {
                    if cameraPermission == .approved {
                        KioskCameraView(onRead: onRead)
                            .frame(width: scannerSize, height: scannerSize)
                            .clipShape(RoundedRectangle(cornerRadius: 12))
                    } else {
                        RoundedRectangle(cornerRadius: 12)
                            .fill(Color("ColorStone200"))
                            .frame(width: scannerSize, height: scannerSize)
                        ProgressView()
                    }

                    // Border
                    RoundedRectangle(cornerRadius: 12)
                        .stroke(Color("ColorBlue600"), lineWidth: 4)
                        .frame(width: scannerSize, height: scannerSize)

                    // Badge
                    VStack {
                        Spacer()
                        HStack(spacing: 8) {
                            ProgressView()
                                .progressViewStyle(CircularProgressViewStyle(tint: .white))
                                .scaleEffect(0.8)
                            Text("Detecting...")
                                .font(.system(size: 14, weight: .medium))
                                .foregroundColor(.white)
                        }
                        .padding(.horizontal, 16)
                        .padding(.vertical, 8)
                        .background(Color("ColorBlue600"))
                        .cornerRadius(100)
                        .offset(y: 20)
                    }
                    .frame(width: scannerSize, height: scannerSize)
                }
            }

            Spacer()

            // Cancel button
            Button(action: onCancel) {
                Text("Cancel")
                    .font(.system(size: 16, weight: .semibold))
                    .foregroundColor(.black)
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, 12)
                    .overlay(
                        RoundedRectangle(cornerRadius: 8)
                            .stroke(Color("ColorStone300"), lineWidth: 1)
                    )
            }
        }
        .padding(30)
        .background(Color("ColorBase1"))
        .ignoresSafeArea(edges: .bottom)
        .onAppear {
            Task {
                switch AVCaptureDevice.authorizationStatus(for: .video) {
                case .authorized:
                    cameraPermission = .approved
                case .notDetermined:
                    cameraPermission = await AVCaptureDevice.requestAccess(for: .video) ? .approved : .denied
                default:
                    cameraPermission = .denied
                }
            }
        }
        .navigationBarBackButtonHidden(true)
    }
}

// MARK: - Camera View

struct KioskCameraView: View {
    var onRead: (String) -> Void

    @State private var session = AVCaptureSession()
    @State private var qrOutput = AVCaptureMetadataOutput()
    @StateObject private var delegate = AVMetadataObjectScannerDelegate()

    var body: some View {
        GeometryReader { geometry in
            CameraView(
                frameSize: CGSize(width: geometry.size.width, height: geometry.size.height),
                session: $session
            )
        }
        .ignoresSafeArea()
        .onAppear(perform: setupCamera)
        .onDisappear { session.stopRunning() }
        .onChange(of: delegate.scannedCode) { newValue in
            if let code = newValue {
                session.stopRunning()
                delegate.scannedCode = nil
                onRead(code)
            }
        }
    }

    private func setupCamera() {
        do {
            guard let device = AVCaptureDevice.DiscoverySession(
                deviceTypes: [.builtInWideAngleCamera],
                mediaType: .video,
                position: .back
            ).devices.first else { return }

            let input = try AVCaptureDeviceInput(device: device)
            guard session.canAddInput(input), session.canAddOutput(qrOutput) else { return }

            session.beginConfiguration()
            session.addInput(input)
            session.addOutput(qrOutput)
            qrOutput.metadataObjectTypes = [.qr]
            qrOutput.setMetadataObjectsDelegate(delegate, queue: .main)
            session.commitConfiguration()

            DispatchQueue.global(qos: .background).async {
                session.startRunning()
            }
        } catch {
            print("Camera setup error: \(error.localizedDescription)")
        }
    }
}

// MARK: - Reader View

public struct KioskReaderView: View {
    @StateObject var delegate: KioskScanViewDelegate
    @Binding var path: NavigationPath
    var onCancel: () -> Void
    var onStartOver: () -> Void

    init(
        uri: String,
        requestedItems: [String: [String: Bool]],
        trustAnchorRegistry: [String]?,
        onCancel: @escaping () -> Void,
        onStartOver: @escaping () -> Void,
        path: Binding<NavigationPath>
    ) {
        self._delegate = StateObject(
            wrappedValue: KioskScanViewDelegate(
                uri: uri,
                requestedItems: requestedItems,
                trustAnchorRegistry: trustAnchorRegistry
            )
        )
        self.onCancel = onCancel
        self.onStartOver = onStartOver
        self._path = path
    }

    public var body: some View {
        VStack {
            switch self.delegate.state {
            case .initializing, .connecting, .connected:
                LoadingView(
                    loadingText: "Connecting...",
                    cancelButtonLabel: "Cancel",
                    onCancel: { self.cancel() }
                )
            case .sendingRequest, .sentRequest, .receivingResponse:
                LoadingView(
                    loadingText: "Processing...",
                    cancelButtonLabel: "Cancel",
                    onCancel: { self.cancel() }
                )
            case .receivedResponse:
                KioskSuccessView(onStartOver: onStartOver)
            case .mdocDisconnected, .error:
                KioskFailureView(onStartOver: onStartOver)
            case .action(.authorizeBluetoothForApp):
                if let url = URL(string: UIApplication.openSettingsURLString) {
                    Button("Authorize Bluetooth to continue") {
                        UIApplication.shared.open(url)
                    }
                    .buttonStyle(.bordered)
                }
            case .action(.turnOnBluetooth):
                Text("Turn on Bluetooth to continue.")
            }
        }
        .padding(.all, 30)
        .navigationBarBackButtonHidden(true)
    }

    func cancel() {
        self.delegate.cancel()
        self.onCancel()
    }
}

// MARK: - Delegate

class KioskScanViewDelegate: ObservableObject & MdocProximityReader.Delegate {
    @Published var state: MdocProximityReader.State = .initializing
    private var mdocReader: MdocProximityReader?

    init(
        uri: String,
        requestedItems: [String: [String: Bool]],
        trustAnchorRegistry: [String]?
    ) {
        self.mdocReader = MdocProximityReader(
            fromHolderQrCode: uri,
            delegate: self,
            requestedItems: requestedItems,
            trustAnchorRegistry: trustAnchorRegistry,
            l2capUsage: .disableL2CAP
        )
    }

    func cancel() {
        self.mdocReader?.disconnect()
    }

    func connectionState(changedTo: SpruceIDMobileSdk.MdocProximityReader.State) {
        DispatchQueue.main.async {
            self.state = changedTo
        }
    }
}

// MARK: - Header

struct KioskHeader: View {
    var body: some View {
        HStack(spacing: 8) {
            Image("SpruceLogo")
                .resizable()
                .frame(width: 28, height: 28)
            Text("Spruce County")
                .font(.system(size: 24, weight: .bold))
                .foregroundColor(Color("ColorStone950"))
        }
        .padding(.top, 15)
    }
}

// MARK: - Success View

struct KioskSuccessView: View {
    var onStartOver: () -> Void

    var body: some View {
        VStack(spacing: 24) {
            KioskHeader()

            Spacer()

            // Icon
            ZStack {
                Circle()
                    .fill(
                        LinearGradient(
                            colors: [Color("ColorBase50"), Color("ColorBlue200")],
                            startPoint: .top,
                            endPoint: .bottom
                        )
                    )
                    .frame(width: 80, height: 80)
                    .overlay(Circle().stroke(Color("ColorBlue300"), lineWidth: 2))
                    .shadow(color: Color.black.opacity(0.15), radius: 4, x: 0, y: 2)
                Image(systemName: "checkmark")
                    .font(.system(size: 40, weight: .medium))
                    .foregroundColor(Color("ColorBlue600"))
            }

            Text("Welcome!")
                .font(.system(size: 36, weight: .bold))
                .foregroundColor(Color("ColorStone950"))

            Text("Your identity has been successfully verified using your mobile driver's license, and a confirmation has been sent to your email.")
                .font(.system(size: 18))
                .foregroundColor(Color("ColorStone600"))
                .multilineTextAlignment(.center)
                .padding(.horizontal, 20)

            Spacer()
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .onTapGesture { onStartOver() }
    }
}

// MARK: - Failure View

struct KioskFailureView: View {
    var onStartOver: () -> Void

    var body: some View {
        VStack(spacing: 24) {
            KioskHeader()

            Spacer()

            // Icon
            ZStack {
                Circle()
                    .fill(
                        LinearGradient(
                            colors: [Color("ColorRose50"), Color("ColorRose200")],
                            startPoint: .top,
                            endPoint: .bottom
                        )
                    )
                    .frame(width: 80, height: 80)
                    .overlay(Circle().stroke(Color("ColorRose300"), lineWidth: 2))
                    .shadow(color: Color.black.opacity(0.15), radius: 4, x: 0, y: 2)
                Image(systemName: "xmark")
                    .font(.system(size: 40, weight: .medium))
                    .foregroundColor(Color("ColorRose600"))
            }

            Text("Invalid")
                .font(.system(size: 36, weight: .bold))
                .foregroundColor(Color("ColorRose700"))

            Text("Your ID has not been accepted. Please try again or check in with reception for verification.")
                .font(.system(size: 18))
                .foregroundColor(Color("ColorStone600"))
                .multilineTextAlignment(.center)
                .padding(.horizontal, 20)

            Spacer()

            Button(action: onStartOver) {
                HStack {
                    Image(systemName: "arrow.left")
                    Text("Start over")
                }
                .font(.system(size: 16, weight: .medium))
                .foregroundColor(Color("ColorBlue800"))
                .padding(.horizontal, 40)
                .padding(.vertical, 14)
                .background(Color("ColorBlue100"))
                .overlay(
                    RoundedRectangle(cornerRadius: 30)
                        .stroke(Color("ColorBlue400"), lineWidth: 1)
                )
                .cornerRadius(30)
            }

            Spacer().frame(height: 60)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}
